apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceExtension.cs

```diff
@@ -0,0 +1,168 @@
+        /// used with <see cref="AddGoogleTrace"/>,
+        /// </summary>
+        public static void UseGoogleTrace(this IApplicationBuilder app)
+        {
+            GaxPreconditions.CheckNotNull(app, nameof(app));
+            app.UseMiddleware<CloudTraceMiddleware>();
+        }
+
+        /// <summary>
+        /// Adds the needed services for Google Cloud Tracing. Used with <see cref="UseGoogleTrace"/>.
+        /// </summary>
+        /// <param name="projectId">The Google Cloud Platform project ID. Cannot be null.</param>
+        /// <param name="config">Optional trace configuration, if unset the default will be used.</param>
+        /// <param name="clientTask">Optional a task which produces the Trace client, if 
+        ///     unset the default will be used.</param>
+        public static void AddGoogleTrace(
+            this IServiceCollection services, string projectId,
+            TraceConfiguration config = null, Task<TraceServiceClient> clientTask = null)
+        {
+            GaxPreconditions.CheckNotNull(services, nameof(services));
+            GaxPreconditions.CheckNotNull(projectId, nameof(projectId));
+
+            clientTask = clientTask ?? TraceServiceClient.CreateAsync();
+            config = config ?? TraceConfiguration.Create();
+
+            IConsumer<TraceProto> consumer = ConsumerFactory<TraceProto>.GetConsumer(
+                 new GrpcTraceConsumer(clientTask), TraceSizer.Instance, config.BufferOptions);
+
+            services.AddSingleton(new ProjectId { Id = projectId });
+            services.AddSingleton(consumer);
+            services.AddSingleton(TraceIdFactory.Create());
+            services.AddSingleton(RateLimitingTraceOptionsFactory.Create(config));
+            services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
+
+            services.AddScoped(CreateTraceHeaderContext);
+            services.AddScoped(CreateManagedTracer);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="TraceHeaderContext"/> based on the current <see cref="HttpContext"/>.
+        /// </summary>
+        internal static TraceHeaderContext CreateTraceHeaderContext(IServiceProvider provider)
+        {
+            var accessor = provider.GetService<IHttpContextAccessor>();
+            string header = accessor?.HttpContext?.Request?.Headers[TraceHeaderContext.TraceHeader];
+            return TraceHeaderContext.FromHeader(header);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="IManagedTracer"/> based on the <see cref="TraceHeaderContext"/> and
+        /// the rate limiter.
+        /// </summary>
+        internal static IManagedTracer CreateManagedTracer(IServiceProvider provider)
+        {
+            // If the trace header says to trace or if the rate limiter allows tracing continue.
+            var headerContext = provider.GetService<TraceHeaderContext>();
+            if (!headerContext.ShouldTrace)
```

Comment:
I added a check to ensure none of them are null.  Unless someone overrides one of the services I don't see this happening, but can't hurt to be safe.